polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -197,186 +218,57 @@ pub fn worker_entrypoint(
-						// and time, it is okay to do extra checks here. This takes negligible time
-						// anyway.
-						if let PrepareJobKind::Prechecking = prepare_job_kind {
-							result = result.and_then(|output| {
-								runtime_construction_check(
-									output.0.as_ref(),
-									executor_params.as_ref(),
-								)?;
-								Ok(output)
-							});
-						}
-
-						result
-					},
-					Arc::clone(&condvar),
-					WaitOutcome::Finished,
-				)?;
-
-				let outcome = thread::wait_for_threads(condvar);
-
-				let peak_alloc = {
-					let peak = end_memory_tracking();
-					gum::debug!(
-						target: LOG_TARGET,
-						%worker_pid,
-						"prepare job peak allocation is {} bytes",
-						peak,
-					);
-					peak
 				};
 
-				let result = match outcome {
-					WaitOutcome::Finished => {
-						let _ = cpu_time_monitor_tx.send(());
-
-						match prepare_thread.join().unwrap_or_else(|err| {
-							Err(PrepareError::Panic(stringify_panic_payload(err)))
-						}) {
-							Err(err) => {
-								// Serialized error will be written into the socket.
-								Err(err)
-							},
-							Ok(ok) => {
-								cfg_if::cfg_if! {
-									if #[cfg(target_os = "linux")] {
-										let (artifact, cpu_time_elapsed, max_rss) = ok;
-									} else {
-										let (artifact, cpu_time_elapsed) = ok;
-									}
-								}
-
-								// Stop the memory stats worker and get its observed memory stats.
-								#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
-								let memory_tracker_stats = get_memory_tracker_loop_stats(memory_tracker_thread, worker_pid);
-								let memory_stats = MemoryStats {
-									#[cfg(any(
-										target_os = "linux",
-										feature = "jemalloc-allocator"
-									))]
-									memory_tracker_stats,
-									#[cfg(target_os = "linux")]
-									max_rss: extract_max_rss_stat(max_rss, worker_pid),
-									// Negative peak allocation values are legit; they are narrow
-									// corner cases and shouldn't affect overall statistics
-									// significantly
-									peak_tracked_alloc: if peak_alloc > 0 {
-										peak_alloc as u64
-									} else {
-										0u64
-									},
-								};
-
-								// Write the serialized artifact into a temp file.
-								//
-								// PVF host only keeps artifacts statuses in its memory,
-								// successfully compiled code gets stored on the disk (and
-								// consequently deserialized by execute-workers). The prepare worker
-								// is only required to send `Ok` to the pool to indicate the
-								// success.
-
-								gum::debug!(
-									target: LOG_TARGET,
-									%worker_pid,
-									"worker: writing artifact to {}",
-									temp_artifact_dest.display(),
-								);
-								fs::write(&temp_artifact_dest, &artifact)?;
-
-								Ok(PrepareStats { cpu_time_elapsed, memory_stats })
-							},
-						}
+				// SAFETY: new process is spawned within a single threaded process
```

Comment:
I guess we could be able to check number of threads and do an assert if >1?

Reply:
Great idea, but there's not a portable way to do it. Looks like there's a `num_threads` crate, but I'd rather not pull in a whole dependency for a single assert. I mean, it'd be nice to verify this property in case someone changes it in the future, but the Linux method of checking number of threads relies on [`proc/` filesystem](https://stackoverflow.com/a/7216699) which we don't have access to after pivot_root. So I'm afraid that by trying to do this we may just introduce more problems. Sorry. 🤷 Let me know if you think this is worth pursuing.

Reply:
Yeah: https://github.com/jhpratt/num_threads/blob/main/src/linux.rs#L5

We'd have to remount `/proc` in the new root filesystem. Not sure it's worth exposing that since it could be a security risk. It's already possible to verify by reading the code that there are no threads, async runtimes, etc.

I guess we can do the remount only in `debug` mode, and have this be a `debug_assert`... WDYT?

Reply:
Let's at least have documentation in all the right places. E.g. top-level (main) should already clearly state that this has to be multi threaded. Would it be possible to have a test that confirms single threadedness? E.g. spawn the process and inspect?

Reply:
You cannot remount /proc without root privilege anyway iiuc.
And `procfs` is not guaranteed to be there either.
The only reliable way to count threads is to intercept `pthread_create`. An example may be found [here](https://github.com/EmbarkStudios/crash-handling/blob/main/crash-handler/src/unix/pthread_interpose.rs). I definitely do not recommend using this technique, it doesn't pay off. But I wonder how much effort it is not to intercept but to completely bury `pthread_create`. We don't need to count threads; we just need to `abort` if anyone tries to create a thread. Maybe search around the resource limits topic? And cgroups as well. There should be a way to limit thread creation.

Reply:
A general thought not connected directly with this PR: the more we move in the security enforcement direction, the more crucial it becomes to have root privileges. Not at full node runtime, of course, but it can start as root, do security things, and then drop privileges before processing any potentially malicious data. Maybe we could start evaluating that.

Reply:
> Let's at least have documentation in all the right places. E.g. top-level (main) should already clearly state that this has to be single-threaded. Would it be possible to have a test that confirms single threadedness? E.g. spawn the process and inspect?

Ah yeah, that is possible. And I will add some more documentation.

> You cannot remount /proc without root privilege anyway iiuc.

We do have root privileges, see [`CLONE_NEWUSER`](https://man7.org/linux/man-pages/man2/unshare.2.html). It's what allows us to do [`pivot_root`](https://man7.org/linux/man-pages/man2/pivot_root.2.html).

> A general thought not connected directly with this PR: the more we move in the security enforcement direction, the more crucial it becomes to have root privileges. Not at full node runtime, of course, but it can start as root, do security things, and then drop privileges before processing any potentially malicious data. Maybe we could start evaluating that.

You mean running under a root user? It's interesting, but opens up a lot of security problems, so we advise people not to do that.

Reply:
> You mean running under a root user?

I mean, doing that like most privileged UNIX daemons do. Starting up as root (or with suid bit), doing things requiring root (open privileged ports, subscribe to audit events, etc.), and then doing `setuid()` to drop privileges to the regular user level.

Reply:
Btw, right now node could drop privileges if it finds out it is running as root, it's a good security measure by itself.

Reply:
Oh I see, that would help with https://github.com/paritytech/polkadot-sdk/issues/2080#issuecomment-1805489768. Worth raising an issue for discussion.

Reply:
Added tests to check num_threads. Also updated the impl guide!

Reply:
Raised follow-ups: https://github.com/paritytech/polkadot-sdk/issues/2323 / https://github.com/paritytech/polkadot-sdk/issues/2324